base/abstractarray.jl

```diff
@@ -23,14 +23,44 @@ end
 
 size{T,N}(t::AbstractArray{T,N}, d) = d <= N ? size(t)[d] : 1
 size{N}(x, d1::Integer, d2::Integer, dx::Vararg{Integer, N}) = (size(x, d1), size(x, d2), ntuple(k->size(x, dx[k]), Val{N})...)
+
+# trait to indicate "vetted" usage for indexing that may not start at 1
+abstract IndicesSafety
+immutable SafeIndices <: IndicesSafety end
+immutable UnsafeIndices <: IndicesSafety end
+
+macro safeindices(ex)
+    esc(_safeindices(ex))
+end
+
+function _safeindices(ex::Expr)
+    if ex.head == :call
+        f = ex.args[1]
+        if f == :size || f == :length
+            return Expr(:call, f, :(Base.SafeIndices()), ex.args[2:end]...)
+        end
+    end
+    return Expr(ex.head, map(_safeindices, ex.args)...)
+end
+
+_safeindices(arg) = arg
+
+# The default is that size is safe, but array types that use non-1
+# indexing should specialize this to make it unsafe without
+# SafeIndices().
+size( ::IndicesSafety, A::AbstractArray)    = size(A)
+size( ::IndicesSafety, A::AbstractArray, d) = size(A,d)  # fixme
+# size(s::IndicesSafety, A::AbstractArray, d) = d <= ndims(A) ? size(s, A)[d] : 1
```

Comment:
For some reason, uncommenting this line causes inference to bail during bootstrap. (Avoiding `ndims` and using the parameter variant, `A::AbstractArray{T,N}`, does not fix the problem either.)

As a consequence, people who write packages for unconventionally-indexed arrays have to specialize both `size(s, A)` and `size(s, A, d)`, which is a little unfortunate but not too terrible.